### ISSUE TITLE
mlflow.tensorflow.autolog() incorrect early stopping check

### ISSUE THREAD

**Issue by ramji-c** [mlflow/mlflow#2831]
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [x] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04.6 LTS
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 1.8.0
- **Python version**: 3.6.10
- **npm version, if running the dev UI**: NA
- **Exact command to reproduce**: mlflow.tensorflow.autolog() with early stopping callback to model.fit(); and set "restore_best_weights" = True

### Describe the problem
When using an early stopping callback with `restore_best_weights=True`, autolog() erroneously assumes early stopping has happened even if training proceeds till last epoch. It computes `restored_epoch = stopped_epoch - max(1, patience)` which produces incorrect epoch number if there was no early stopping. There should be an additional check to see if `stopped_epoch != n_epochs`, to ensure early stopping has happened. 
This has the side effect of flipping metric plot in mlflow UI, putting an older metric value at a higher step value, giving the impression of increasing loss or decreasing accuracy.

### Code to reproduce issue
Any tf.keras.Model code(v2.1 or higher), with this callback added to `model.fit()` 
```
tf.keras.callbacks.EarlyStopping(monitor='val_loss',
                                                       min_delta=0.01,
                                                       patience=2,
                                                       verbose=0,
                                                       mode='auto',
                                                       baseline=None,
                                                       restore_best_weights=True
                                                       )
```

### Other info / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

**Comment by smurching** [comment]
Hi @ramji-c, thanks for reporting this issue! Do you have an example code snippet that I can run to reproduce it? We have a test [here](https://github.com/mlflow/mlflow/blob/c8ec506e83e7f7e3e0af6094444d7969ed1a11e3/tests/tensorflow_autolog/test_tensorflow2_autolog.py#L230-L249) that [fits a model for 10 epochs](https://github.com/mlflow/mlflow/blob/c8ec506e83e7f7e3e0af6094444d7969ed1a11e3/tests/tensorflow_autolog/test_tensorflow2_autolog.py#L174-L193) with a patience of 11, so that early stopping cannot occur, and verifies that autologging doesn't log an extra metric value in that case In particular, only 10 metric values are logged, one per step of training, rather than 11 metrics

I updated the autologging test in [this branch](https://github.com/smurching/mlflow/pull/new/tensorflow-repro) to use a patience value of 2 and `min_delta` of 0.01, as in your example callback, and found that it still passes using TensorFlow 2.2.0. It's possible there's something different about the test setup vs in your use case - having a script handy that I can run reproduce the error you saw would help rule that out. Thanks!

As you pointed out, we only log metric values again if the Keras callback has a non-zero `stopped_epoch`. If `stopped_epoch == num_epochs`, this actually means that loss has improved by < `min_delta` for the last `patience` epochs, as per the [tf.Keras docs](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping#arguments_1), so early stopping is in fact being applied. For example, if I train a model for 10 epochs with `patience=2` and `min_delta=0.01` and find that early stopping occurs on epoch 10, that implies that loss has improved by less than 0.01 for both epochs 9 and 10, so we should apply early stopping and restore weights from epoch 8.

**Comment by ramji-c** [comment]
Had a brief discussion with Sid and got all clarifications I needed. This can be closed! Thank you, Sid.
